### DISCUSSION TITLE
Test mode settings

### DISCUSSION THREAD

**Discussion by benjamind2008** [revvv/gdash-export-CrLi#8] [Ideas]
With the oldest GDash from 2013 the test mode would play the game with the graphics features turned on. It seems as though the test mode of this new version of GDash does not use the full graphics features and looks like a "default" graphics mode. It's still every bit as smooth and enjoyable as the full graphic rendering, but looks very different to the old 2013 test mode that I remember. It looks as though a rendering feature hasn't been enabled.

All up this is a fantastic upgrade, and the white flash when collecting the last required diamond is definitely something I've wanted to see for a long time. It's also more stable on my system. However, the scrolling isn't as smooth on the 144Hz and 240Hz modes on my LG 27" OLED monitor, but it is perfect in both 60Hz and 120Hz. Unsure of what is happening here, my system is a 3900x Ryzen 9 with a WX3100 GPU, though any GPU I presume should be more than fast enough for this.

**Comment by revvv** [comment]
> It seems as though the test mode of this new version of GDash does not use the full graphics features and looks like a "default" graphics mode.

@cwscws already reported this problem. I'm not sure what caused this issue. I'll check that.

> ... the white flash when collecting the last required diamond is definitely something I've wanted to see for a long time. It's also more stable on my system.

Thanks, but it wasn't me ;-) I just took over Zoltan's latest code...

> However, the scrolling isn't as smooth on the 144Hz and 240Hz modes on my LG 27" OLED monitor, but it is perfect in both 60Hz and 120Hz. Unsure of what is happening here, my system is a 3900x Ryzen 9 with a WX3100 GPU, though any GPU I presume should be more than fast enough for this.

I simply don't know...

**Comment by revvv** [comment]
Could you describe how to distinct between the chosen modes?
For me it looks OK on Windows. I simply see the title of the test window is "gdash.exe" (GTK) instead of "GDash" (SDL/OpenGL) and the mouse cursor is visible (GTK) instead of not visible (SDL/OpenGL).

**Comment by cwscws** [comment]
Simpy use a shader for OpenGL - then you can see the difference between the two modes

**Comment by revvv** [comment]
Looks like that the test game in OpenGL mode was removed by intention: [mainwindow.cpp](https://github.com/revvv/gdash-export-CrLi/blob/b794dbe10c7869b05bd18af62df6a8802bc63af3/src/mainwindow.cpp#L90)

I could reactivate the feature successfully on Windows, but I guess I have to test Mac/Linux as well before I check it in.

**Reply by benjamind2008** [reply]
How does the OpenGL work on Mac? I plan to get a Mac at some point in the distant future, and it would be lovely to have all those nice shaders and silky smooth scrolling on Mac if this is possible (it should be).

**Reply by cwscws** [reply]
OpenGL is deprecated on MacOS as Apple is preferring their own graphics interface Metal. So it's stuck at version 4.1 and not being updated. A game like GDash doesn't need the newer functions so basic functionality is always provided. But you never know if Apple removes OpenGL at some point in the future. So I think it would be not wise to rely on OpenGL that much.

It would be wiser to implement Metal support for the Mac version but I do not know how complex this would be. In other programs there are efforts to add bgfx (from MAME) and lately Vulcan (as there is a translation layer MoltenVK available).

**Comment by cwscws** [comment]
After thinking about this twice I came to the conclusion that it is indeed intentional: The editor is running in GTK+ so the test cave must also be in GTK+ to be consistent. The problem is this: if using OpenGL for the test game and there are troubles with OpenGL you can never ever test caves. The game itself can be set to GTK+, SDL or OpenGL. But if you change the test cave option to always run in OpenGL this could be a potential problem.

In my opinion the only good solution would be an option in the editor preferences like "Use OpenGL for Test Cave" ("OpenGL für Teste Höhle verwenden"). When it's activated it uses OpenGL, if it's deactivated it uses GTK+.

If this is possible. If not, it would be better to leave it as it is.

**Reply by benjamind2008** [reply]
Is it possible to have an option in the editor that allows it to run in OpenGL for the editing as well? Since it would be rather nice to have the game looking the same in the editor as it would appear when the game is running.

Hopefully this is possible. There are many things that can be done, but graphics is my number one concern, and the better it looks, and the more options that are there for users, the better the experience will be.

**Reply by cwscws** [reply]
I think this does not work because of missing menus and mouse pointer. I still think it would be best to leave it as it is for (future) compatibility reasons. Or add the option in the editor preferences to switch OpenGL in Test Cave on/off.

**Reply by benjamind2008** [reply]
Good point. Letting the user decide whether or not to use OpenGL during a test cave would be sufficient enough.

**Comment by revvv** [comment]
IMHO if you lock yourself out with OpenGL, you have to delete `gdash.ini'. So I don't think a new settings is necessary.
For now I release a beta which enables OpenGL for the test game.

**Reply by cwscws** [reply]
I think there should be an option for that. A specific option for the editor.

When deleting the .ini file, the test cave uses GTK+ again?

**Reply by revvv** [reply]
Yes, that's the default. If you want to keep your settings, just edit the .ini file:

    graphics_engine=0

0=GTK, 1=SDL, 2=OpenGL

If you lock yourself out, you can restart GDash, then
- Press O for options. Configure engine _or_
- Press E for edit. Then use GTK menu to configure engine.

IMHO there is no need for an additional option. And BTW I never locked myself out.
If you give me feedback to the current pre-release I'll check it in.

**Reply by cwscws** [reply]
It seems to be ok.

Tested it with switching to GTK+ and the Test Cave used GTK+ again. Switching to OpenGL and Test Cave uses OpenGL.
So it's working!

Btw: I noticed that with GTK+ caves with many things going on the screen it slows down a lot - at least at 5 x size. No slowdown with OpenGL. As my computer is quite fast I wonder why the slowdown in GTK+.
